nibabel/loadsave.py

```diff
@@ -106,25 +58,38 @@ def save(img, filename):
     -------
     None
     '''
+
+    # Save the type as expected
     try:
         img.to_filename(filename)
     except ImageFileError:
         pass
     else:
         return
-    froot, ext, trailing = splitext_addext(filename, ('.gz', '.bz2'))
+
+    # Be nice to users by making common implicit conversions
+    froot, ext, trailing = splitext_addext(filename, img._compressed_exts)
+    lext = ext.lower()
+
     # Special-case Nifti singles and Pairs
-    if type(img) == Nifti1Image and ext in ('.img', '.hdr'):
+    from .nifti1 import Nifti1Image, Nifti1Pair  # Inline imports, as this module
+    from .nifti2 import Nifti2Image, Nifti2Pair  # really shouldn't reference any image type
+    if type(img) == Nifti1Image and lext in ('.img', '.hdr'):
         klass = Nifti1Pair
-    elif type(img) == Nifti2Image and ext in ('.img', '.hdr'):
+    elif type(img) == Nifti2Image and lext in ('.img', '.hdr'):
         klass = Nifti2Pair
-    elif type(img) == Nifti1Pair and ext == '.nii':
+    elif type(img) == Nifti1Pair and lext == '.nii':
         klass = Nifti1Image
-    elif type(img) == Nifti2Pair and ext == '.nii':
+    elif type(img) == Nifti2Pair and lext == '.nii':
         klass = Nifti2Image
-    else:
-        img_type = ext_map[ext]
-        klass = class_map[img_type]['class']
+    else:  # arbitrary conversion
+        valid_klasses = filter(lambda klass: klass.is_valid_extension(ext),
+                               all_image_classes)
+        try:
+            klass = next(iter(valid_klasses))
+        except StopIteration:  # if iterator is empty
+            raise ImageFileError('Cannot work out file type of "%s"' %
+                                 filename)
```

Comment:
The `next(iter(filter(...)))` is awkward. Alternatives:

    valid_klasses = list(filter(lambda klass: klass.is_valid_extension(ext), all_image_classes))
    valid_klasses = [klass for klass in all_image_classes if klass.is_valid_extension(ext)]